core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -273,7 +273,8 @@ abstract class AbstractFetcherThread(name: String,
 
                     logAppendInfoOpt.foreach { logAppendInfo =>
                       val nextOffset = logAppendInfo.lastOffset + 1
-                      fetcherLagStats.getAndMaybePut(topicPartition).lag = Math.max(0L, partitionData.highWatermark - nextOffset)
+                      fetcherLagStats.getAndMaybePut(topicPartition).lag =
+                        if (nextOffset == 0) 0 else Math.max(0L, partitionData.highWatermark - nextOffset)
```

Comment:
@huxihx : Thanks for finding the issue. We probably can't just make lag 0 if nextOffset is 0. It's possible for the follower to get an empty data set (e.g. the replication is throttled) but a new high watermark. In that case, the lag may not be 0. Perhaps we can fix the issue like the following.

```
                      val nextOffset = if (logAppendInfo.validBytes > 0) logAppendInfo.lastOffset + 1 else currentFetchState.fetchOffset
                      fetcherLagStats.getAndMaybePut(topicPartition).lag = Math.max(0L, partitionData.highWatermark - nextOffset)
```

Reply:
Thanks for the comments. Yes, it's more much reasonable to use the current fetch offset instead of zero when an empty data set is retrieved from the leader.